patches/server/1030-PreSpawnerSearchEvent.patch

```diff
@@ -0,0 +1,57 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: TreemanKing <67459602+TreemanKing@users.noreply.github.com>
+Date: Wed, 16 Aug 2023 11:53:44 +1000
+Subject: [PATCH] PreSpawnerSearchEvent
+
+This patch introduces a new event, PreSpawnerSearchEvent. This event is designed to be fired before a mob spawner attempts to search for nearby players to activate. It provides the ability to control and customize the activation logic of mob spawners, preventing them from searching and spawning mobs if desired
+
+diff --git a/src/main/java/net/minecraft/world/level/BaseSpawner.java b/src/main/java/net/minecraft/world/level/BaseSpawner.java
+index 633500aefd515df5dadda3802b94079f75a03fa0..9c3ee094035e6d903d3732349c9ff1e0a93b891a 100644
+--- a/src/main/java/net/minecraft/world/level/BaseSpawner.java
++++ b/src/main/java/net/minecraft/world/level/BaseSpawner.java
+@@ -46,6 +46,7 @@ public abstract class BaseSpawner {
+     public int requiredPlayerRange = 16;
+     public int spawnRange = 4;
+     private int tickDelay = 0; // Paper
++    private int searchCooldownTicks = 0; // Paper
+ 
+     public BaseSpawner() {}
+ 
+@@ -54,7 +55,7 @@ public abstract class BaseSpawner {
+         this.spawnPotentials = SimpleWeightedRandomList.empty(); // CraftBukkit - SPIGOT-3496, MC-92282
+     }
+ 
+-    public boolean isNearPlayer(Level world, BlockPos pos) {
++    public boolean isNearPlayer(Level world, BlockPos pos) { // diff on change - PreSpawnerSearchEvent
+         return world.hasNearbyAlivePlayerThatAffectsSpawning((double) pos.getX() + 0.5D, (double) pos.getY() + 0.5D, (double) pos.getZ() + 0.5D, (double) this.requiredPlayerRange); // Paper - Affects Spawning API
+     }
+ 
+@@ -86,7 +87,27 @@ public abstract class BaseSpawner {
+         tickDelay = world.paperConfig().tickRates.mobSpawner;
+         if (tickDelay == -1) { return; } // If disabled
+         // Paper end
+-        if (this.isNearPlayer(world, pos)) {
++
++        // Paper start - PreSpawnerSearchEvent
++        if (searchCooldownTicks > 0) {
++            --searchCooldownTicks;
++            return;
++        }
++
++        java.util.Collection<org.bukkit.entity.Player> activatingPlayers = com.google.common.collect.Collections2.transform(world.getNearbyPlayers(null, (double) pos.getX() + 0.5D, (double) pos.getY() + 0.5D, (double) pos.getZ() + 0.5D, (double) this.requiredPlayerRange, net.minecraft.world.entity.EntitySelector.PLAYER_AFFECTS_SPAWNING), net.minecraft.server.level.ServerPlayer::getBukkitEntity); // Uses logic from getNearbyPlayer
++
++        io.papermc.paper.event.entity.PreSpawnerSearchEvent preSpawnerSearchEvent = new io.papermc.paper.event.entity.PreSpawnerSearchEvent(io.papermc.paper.util.MCUtil.toLocation(world, pos), activatingPlayers, Math.max(0, searchCooldownTicks));
++
++        if (!preSpawnerSearchEvent.callEvent()) return;
```

Comment:
The event is not anymore cancellable with this way so this can just be a simple #callEvent